### PR TITLE
Support limits for silences

### PR DESCRIPTION
**Pull request by grobinson-grafana**
#### What this PR does

This pull request adds support for silence limits https://github.com/prometheus/alertmanager/pull/3852.

#### Which issue(s) this PR fixes or relates to

Fixes #<issue number>

#### Checklist

- [x] Tests updated.
- [x] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [x] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/util/validation/limits.go [hunk 1]
```diff
@@ -335,6 +337,8 @@ func (l *Limits) RegisterFlags(f *flag.FlagSet) {
 	}
 	f.Var(&l.NotificationRateLimitPerIntegration, "alertmanager.notification-rate-limit-per-integration", "Per-integration notification rate limits. Value is a map, where each key is integration name and value is a rate-limit (float). On command line, this map is given in JSON format. Rate limit has the same meaning as -alertmanager.notification-rate-limit, but only applies for specific integration. Allowed integration names: "+strings.Join(allowedIntegrationNames, ", ")+".")
 	f.IntVar(&l.AlertmanagerMaxConfigSizeBytes, "alertmanager.max-config-size-bytes", 0, "Maximum size of configuration file for Alertmanager that tenant can upload via Alertmanager API. 0 = no limit.")
+	f.IntVar(&l.AlertmanagerMaxSilences, "alertmanager.max-silences", 0, "Maximum number of active and pending silences that a tenant can have at once. 0 = no limit.")
+	f.IntVar(&l.AlertmanagerMaxPerSilenceSizeBytes, "alertmanager.max-per-silence-size-bytes", 0, "Maximum per silence size. 0 = no limit.")
```
Comment (alexweav) [inline pkg/util/validation/limits.go:341]: Enormous nitpick: `max-per-silence-size-bytes`->`max-silence-size-bytes`
It's shorter and consistent with `max-template-size-bytes` (applies to wherever the term appears)

File: pkg/util/validation/limits.go [hunk 2]
```diff
@@ -335,6 +337,8 @@ func (l *Limits) RegisterFlags(f *flag.FlagSet) {
 	}
 	f.Var(&l.NotificationRateLimitPerIntegration, "alertmanager.notification-rate-limit-per-integration", "Per-integration notification rate limits. Value is a map, where each key is integration name and value is a rate-limit (float). On command line, this map is given in JSON format. Rate limit has the same meaning as -alertmanager.notification-rate-limit, but only applies for specific integration. Allowed integration names: "+strings.Join(allowedIntegrationNames, ", ")+".")
 	f.IntVar(&l.AlertmanagerMaxConfigSizeBytes, "alertmanager.max-config-size-bytes", 0, "Maximum size of configuration file for Alertmanager that tenant can upload via Alertmanager API. 0 = no limit.")
+	f.IntVar(&l.AlertmanagerMaxSilences, "alertmanager.max-silences", 0, "Maximum number of active and pending silences that a tenant can have at once. 0 = no limit.")
```
Comment (alexweav) [inline pkg/util/validation/limits.go:340]: Similar naming nitpick: `max-silences` -> `max-silences-count` for consistency with `max-templates-count` and `max-alerts-count`

**Review by alexweav — APPROVED** [review]

File: pkg/util/validation/limits.go (hunk 1 above)
Comment (grobinson-grafana) [reply pkg/util/validation/limits.go]: Ha! You suggested it at the same time I fixed it 🙂 I changed it to make it the same as the others!

File: pkg/alertmanager/alertmanager.go [hunk 3]
```diff
@@ -220,8 +220,12 @@ func New(cfg *Config, reg *prometheus.Registry) (*Alertmanager, error) {
 	am.silences, err = silence.New(silence.Options{
 		SnapshotFile: silencesFile,
 		Retention:    cfg.Retention,
-		Logger:       log.With(am.logger, "component", "silences"),
-		Metrics:      am.registry,
+		Limits: silence.Limits{
+			MaxSilences:        cfg.Limits.AlertmanagerMaxSilencesCount(cfg.UserID),
+			MaxPerSilenceBytes: cfg.Limits.AlertmanagerMaxSilenceSizeBytes(cfg.UserID),
```
Comment (56quarters) [inline pkg/alertmanager/alertmanager.go:225]: `max-per-silence-bytes` makes this limit a lot easier to understand (compared to `max-silence-size-bytes`). WDYT about using max _per_ silence bytes everywhere? OK too if you'd like to keep it as-is for consistency with other limits.

**Review by 56quarters — APPROVED** [review]
LGTM with one non-blocking comment

File: pkg/alertmanager/alertmanager.go (hunk 3 above)
Comment (grobinson-grafana) [reply pkg/alertmanager/alertmanager.go:225]: I would like to keep it consistent with other limits 🙂

**Review by alexweav — APPROVED** [review]

**Pull request merged**